api/v1beta1/clusteripamclaim_types.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"errors"
+	"net"
+	"net/netip"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ClusterIPAMClaimKind Denotes the clusteripamclaim resource Kind.
+	ClusterIPAMClaimKind = "ClusterIPAMClaim"
+
+	// InvalidClaimConditionType Denotes that the claim is invalid
+	InvalidClaimConditionType = "InvalidClaimCondition"
+)
+
+// ClusterIPAMClaimSpec defines the desired state of ClusterIPAMClaim
+type ClusterIPAMClaimSpec struct {
+	// Provider is the name of the provider that this claim will be consumed by
+	Provider string `json:"provider,omitempty"`
+
+	// NodeNetwork defines the allocation requisitioning ip addresses for cluster nodes
+	NodeNetwork AddressSpaceSpec `json:"nodeNetwork,omitempty"`
+
+	// ClusterNetwork defines the allocation for requisitioning ip addresses for use by the k8s cluster itself
+	ClusterNetwork AddressSpaceSpec `json:"clusterNetwork,omitempty"`
+
+	// ExternalNetwork defines the allocation for requisitioning ip addresses for use by services such as load balancers
+	ExternalNetwork AddressSpaceSpec `json:"externalNetwork,omitempty"`
+
+	// Cluster is the reference to the cluster deployment that this claim is for
+	Cluster string `json:"cluster"`
+
+	// ClusterIPAMRef is the reference to the [ClusterIPAM] resource that this claim is for
+	ClusterIPAMRef string `json:"clusterIPAMRef,omitempty"`
+}
+
+// AddressSpaceSpec defines the ip address space that will be allocated
+type AddressSpaceSpec struct {
+	// CIDR notation of the allocated address space
+	CIDR string `json:"cidr,omitempty"`
+
+	// IPAddresses to be allocated
+	IPAddresses []string `json:"ipAddresses,omitempty"`
+}
+
+// ClusterIPAMClaimStatus defines the observed state of ClusterIPAMClaim
+type ClusterIPAMClaimStatus struct {
+	// +kubebuilder:default:=false
+
+	// Bound is a flag to indicate that the claim is bound because all ip addresses are allocated
+	Bound bool `json:"bound"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
```

Comment:
```suggestion
type ClusterIPAMClaimStatus struct {
	// +patchMergeKey=type
	// +patchStrategy=merge
	// +listType=map
	// +listMapKey=type

	// TODO: meaningful comment
	Conditions []metav1.Condition `json:"conditions,omitempty"`
	// +kubebuilder:default:=false

	// Bound is a flag to indicate that the claim is bound because all ip addresses are allocated
	Bound bool `json:"bound"`
}
```